packages/nextjs/src/common/wrapRouteHandlerWithSentry.ts

```diff
@@ -0,0 +1,80 @@
+import { addTracingExtensions, captureException, flush, getCurrentHub, runWithAsyncContext, trace } from '@sentry/core';
+import { addExceptionMechanism, tracingContextFromHeaders } from '@sentry/utils';
+
+import type { RouteHandlerContext } from './types';
+import { platformSupportsStreaming } from './utils/platformSupportsStreaming';
+
+/**
+ * Wraps a Next.js route handler with performance and error instrumentation.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function wrapRouteHandlerWithSentry<F extends (...args: any[]) => any>(
+  routeHandler: F,
+  context: RouteHandlerContext,
+): (...args: Parameters<F>) => ReturnType<F> extends Promise<unknown> ? ReturnType<F> : Promise<ReturnType<F>> {
+  addTracingExtensions();
+
+  const { method, parameterizedRoute, baggageHeader, sentryTraceHeader } = context;
+
+  return new Proxy(routeHandler, {
+    apply: (originalFunction, thisArg, args) => {
+      return runWithAsyncContext(async () => {
+        const hub = getCurrentHub();
+        const currentScope = hub.getScope();
+
+        const { traceparentData, dynamicSamplingContext, propagationContext } = tracingContextFromHeaders(
+          sentryTraceHeader,
+          baggageHeader,
+        );
+        currentScope.setPropagationContext(propagationContext);
+
+        let res;
+        try {
+          res = await trace(
+            {
+              op: 'http.server',
+              name: `${method} ${parameterizedRoute}`,
+              status: 'ok',
+              ...traceparentData,
+              metadata: {
+                source: 'route',
+                dynamicSamplingContext: traceparentData && !dynamicSamplingContext ? {} : dynamicSamplingContext,
+              },
+            },
+            async span => {
+              const response: Response = await originalFunction.apply(thisArg, args);
+
+              try {
+                span?.setHttpStatus(response.status);
+              } catch {
+                // best effort
+              }
+
+              return response;
+            },
+            error => {
+              captureException(error, scope => {
+                scope.addEventProcessor(event => {
+                  addExceptionMechanism(event, {
+                    handled: false,
+                  });
+                  return event;
+                });
+
+                return scope;
```

Comment:
l: do we need to filter specific errors here (http errors/redirects)? Just asking because we do this in SvelteKit. Feel free to disregard if not applicable.

Reply:
Very good point! Next.js throws an error when calling `redirect()` so we need to handle that. [ea87659](https://github.com/getsentry/sentry-javascript/pull/8832/commits/ea8765916f41bb27444d967dcd597b1589a57892)